### PR TITLE
Adds enhancements for the DPC program model

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

Adds enhancements for the DPC program model.

## Changes to Accounts

- Adds `Copy` trait to `Address` struct

## Changes to Records

- Removes `position` from `Record` struct (it is never used)
- Adds `Record::new_noop_input()` and `Record::new_noop_output()`
- Adds `Record::new_input()` and `Record::new_output()`

## Changes to Programs

- Includes #343 
- Reverses program public variables declaration to match allocation order
- Changes program circuits to be referenced from `circuit_id` instead of `circuit_index` (error-prone convention)
- Adds `Parameters::noop_circuit_id()` only for use in `load()` methods
- Adds helper method `fn program_circuit_id(verifying_key: &<Self::ProgramSNARK as SNARK>::VerifyingKey) -> Result<Self::ProgramCircuitID>` into `Parameters`
- Renames `Parameters::program_id_crh()` to `Parameters::program_circuit_id_crh()`
- Renames `Parameters::program_id_tree_parameters()` to `Parameters::program_circuit_tree_parameters()`

## Changes to DPC

- Introduces `Executable` to `DPC::execute()` which leverage trait objects to abstract the program executions
- Adds `DPCError::MissingCircuit`

### CODE REVIEW

File: dpc/src/program/public_variables.rs [hunk 1]
```diff
@@ -15,41 +15,30 @@
...
 #[derive(Derivative)]
 #[derivative(
     Clone(bound = "C: Parameters"),
     Debug(bound = "C: Parameters"),
     Default(bound = "C: Parameters")
 )]
-pub struct ProgramPublicVariables<C: Parameters> {
-    pub local_data_root: C::LocalDataRoot,
+pub struct PublicVariables<C: Parameters> {
     pub record_position: u8,
```
Comment (weikengchen) [inline dpc/src/program/public_variables.rs:27]: Just to confirm: The PR said that we removed the record position from the `Record` struct, but `position` will still be used in other places to reference to this record right?

File: parameters/examples/genesis.rs [hunk 2]
```diff
@@ -58,71 +58,51 @@ pub fn generate<C: Parameters>(recipient: &Address<C>, value: u64) -> Result<(Ve
...
-        recipient.clone(),
+        recipient,
         false,
         value,
         Payload::default(),
         C::NUM_INPUT_RECORDS as u8,
         joint_serial_numbers.clone(),
         rng,
     )?);
-    new_records.push(Record::new_full(
```
Comment (weikengchen) [inline parameters/examples/genesis.rs:91]: I agree the input/output records are better than old/new records, as the old/new records may give people a feeling that we are just updating existing records.

**Review by weikengchen — APPROVED** [review]
The changes look reasonable.

File: dpc/src/program/public_variables.rs (hunk 1 above)
Comment (howardwu) [reply dpc/src/program/public_variables.rs:27]: Correct, there is a discrepancy that we need to resolve (via a standard convention) for the `leaf_index` and `record_position` as both are part of local data. They arose for different reasons, but are ordered in the same manner.

**Pull request merged**